src/StreamFactoryInterface.php

```diff
@@ -9,15 +9,11 @@
     /**
      * Create a new stream from a resource or a string.
      *
-     * If the argument is a resource, it MUST be readable and SHOULD be seekable. It MAY be writable.
-     * If the argument is a string, a temporary resource will be created that is writable and seekable.
-     * 
-     * If the arugment is a string it will be interpided as the content of the stream. File names or 
-     * file paths are not supported. 
+     * String content will be used as the body of the stream. Files paths are not supported.
```

Comment:
> Note that I'm proposing two distinct factories for php://temp and php://memory based streams - one is suitable for smaller chunks, the other more suitable for writing large streams such as data-exports.

Except this is not how `php://temp` works. As per [the docs][1]:

> The only difference between the two is that `php://memory` will always store its data in memory, whereas `php://temp` will use a temporary file once the amount of data stored hits a predefined limit (the default is 2 MB). 

It is perfectly acceptable to **always** rely on `php://temp` without worrying about memory/fs issues.

> the interfaces minimally provide that without being opinionated, dictating implementation details, or making assumptions (or mandates) about the underlying implementations.

Except that this is exactly what PSRs do: dictate how implementations work. PSR-7 is very opinionated (eg: immutability) and is better for it.

> The PSR itself doesn't need to concern itself with ease of use. 

This statement makes no sense to me. Any interface that is suggested to be standard should absolutely consider how the user will interact with it. Your example of `CommonStreamFactory` highlights exactly why the interface should work as I describe here. If users start to rely on a specific implementation (`CommonStreamFactory`) then the interfaces are not serving the users' needs and we are no better off than users relying on a specific implementation.

Ultimately the goal is that users can create streams without relying on a specific implementation. Wether that stream comes from an existing resource or a not is up the user. However, the target audience of this PSR is not so much the end user as it is package and framework authors. To this end, these developers will be able to significantly reduce the amount of "detection" code required by simply assuming that any incoming stream body is either a resource or a string of content, because that is what the interface allows for.

As a middleware author, the burden of determining what kind of content I am dealing with is deferred to the factory implementation. As an end user, I can choose to generate strings of content, or use existing resources if I have more specific needs. This is the best case scenario and everyone has the ability to use what they want without undue burden.

I don't see any use cases that are better served by having a bunch of different interfaces that need to be selected from, as it will only make code using said interfaces more likely to not serve the need that I have in any particular situation. For example, if a particular middleware only used `FileStreamFactory` I would have no recourse except to not use that middleware, or fork the middleware and inject the type of stream I want to use.

From my experience as a package author and package user, the implementation I am suggesting is the right balance of convenience and flexibility. Want to simply send a string of content? Covered. Want to send a file or restrict the stream to be read-only? Covered. Want to only use memory based streams? Covered.

Besides supposed purity, what benefit would having multiple streams have for package authors and end users?

[1]: http://php.net/manual/en/wrappers.php.php#refsect2-wrappers.php-unknown-unknown-unknown-unknown-unknown-descriptios